docs/rfcs/xxxx-river.md

```diff
@@ -0,0 +1,400 @@
+  of Kubernetes Service Discovery into HTTP Service Discovery)
+
+* Allowing users to perform custom conditional logic, such as increasing rate
+  limits during busier business months.
+
+Without expressions, we would need more components for common tasks. A
+`concat()` function call can be used to combine lists of discovered Prometheus
+targets, but without expressions, there would likely need to be a dedicated
+component for aggregating sets of targets together.
+
+The belief is that the work required to use and maintain an expression language
+is far less than the combined work to implement features that would be handled
+by expressions out of the box.
+
+YAML by itself does not support expressions. While expressions could be added
+to YAML through the use of templates (e.g., `field_a: {{ some_variable + 5
+}}`), it is beyond the scope of what YAML was intended for and would be more
+cumbersome to use compared to a language where expressions are a first-class
+concept.
+
+### Why an embedded language?
+
+Embedded languages are typically known for the ability for maintainers of the
+project to expose APIs to users of the embedded language, such as the Lua API
+used by Neovim. Embedded languages typically imply tight integration with the
+application embedding them, as opposed to something like YAML which is a
+language consumed once at load time.
+
+An embedded language is a good fit for Flow:
+
+* It makes it easy for developers to expose APIs which users can interact with
+  or pass around. These APIs can be opaque arbitrary Go types which the user
+  doesn't need to know the detail of, only that it refers to something like a
+  stream of metric samples.
+
+* It is well-suited for continuous evaluation (i.e., the core feature of Flow)
+  so configuration can adapt to a changing environment.
+
+### Why a declarative language? Why not Lua?
+
+The language Flow relies on should have a minimal learning curve. While a
+language like Lua could likely be a decent fit for Flow, imperative languages
+have steeper learning curves compared to declarative languages.
+
+Declarative languages natively map to configuration files, since configuration
+files are used to tell the application the desired state, reducing the learning
+curve for the language and making it easier for users to reason about what the
+final config state should be.
+
+### Why not HCL?
+
+> For some background, it's important to note that HCL can be considered two
+> separate projects: hashicorp/hcl (the language and expression evaluator) and
+> zclconf/go-cty (the value and type system used by HCL).
+
+HCL was the obvious first choice for the Flow prototype: it supports
+expressions, you can expose functions for users to call, and its syntax has a
+small learning curve.
+
+However, I found the [schema-driven processing][] API exposed by HCL to be
+difficult to work with for Flow, requiring a lot of boilerplate. While there is
```

Comment:
Could you elaborate this a bit more? Where would be this boilerplate in the codebase and why is worst that writing a new languange?

Reply:
Added detail:

> While the fork lets us avoid the boilerplate of hand-crafting
schema definitions for components, it contains a non-trivial amount of changes
that would need to be contributed upstream to be tenable long-term.